File: app/src/main/java/com/cazimir/relaxoo/dialog/favorite/SaveToFavoritesDialog.java
package com.cazimir.relaxoo.dialog.favorite;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;

import com.cazimir.relaxoo.R;
import com.cazimir.relaxoo.dialog.RetainableDialogFragment;
import com.cazimir.relaxoo.model.SavedCombo;
import com.cazimir.relaxoo.model.Sound;
import com.cazimir.relaxoo.ui.sound_grid.OnFavoriteSaved;

import java.util.List;

public class SaveToFavoritesDialog extends RetainableDialogFragment
        implements DialogInterface.OnClickListener {

  private static final String TAG = "SaveToFavoritesDialog";
  private final List<Sound> playingSoundIds;
  private OnFavoriteSaved addToListInFragmentCallback;
  private View form = null;

  public SaveToFavoritesDialog(List<Sound> sounds) {
    this.playingSoundIds = sounds;
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {

    addToListInFragmentCallback = (OnFavoriteSaved) getContext();

    form = getActivity().getLayoutInflater().inflate(R.layout.favorites_dialog, null);

    AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

    return (builder
        .setTitle("Save Favorites Combos")
        .setView(form)
        .setPositiveButton(android.R.string.ok, this)
        .setNegativeButton(android.R.string.cancel, null)
        .create());
  }

  @Override
  public void onClick(DialogInterface dialog, int which) {
    EditText comboName = form.findViewById(R.id.comboName);
    // saving a new combo
    SavedCombo savedCombo =
            new SavedCombo.Builder()
                    .withName(comboName.getText().toString())
                    .withSoundPoolParameters(playingSoundIds)
                    .withPlaying(true)
                    .build();
    addToListInFragmentCallback.saved(savedCombo);
    Toast.makeText(getActivity(), getString(R.string.saved_combo), Toast.LENGTH_LONG).show();
  }

  @Override
  public void onDismiss(DialogInterface unused) {
    super.onDismiss(unused);

    Log.d(getClass().getSimpleName(), "Goodbye!");
  }

  @Override
  public void onCancel(DialogInterface unused) {
    super.onCancel(unused);

    Toast.makeText(getActivity(), "Canceled", Toast.LENGTH_LONG).show();
  }
}

File: app/src/main/java/com/cazimir/relaxoo/dialog/favorite/SaveToFavoritesDialog.kt
package com.cazimir.relaxoo.dialog.favorite

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.cazimir.relaxoo.R
import com.cazimir.relaxoo.dialog.RetainableDialogFragment
import com.cazimir.relaxoo.model.SavedCombo
import com.cazimir.relaxoo.model.Sound
import com.cazimir.relaxoo.ui.sound_grid.OnFavoriteSaved
import com.cazimir.utilitieslibrary.onChange
import kotlinx.android.synthetic.main.favorites_dialog.view.*

class SaveToFavoritesDialog(private val playingSoundIds: List<Sound>) : RetainableDialogFragment(), DialogInterface.OnClickListener {

    private lateinit var positiveButton: Button
    private var touched = false
    private var addToListInFragmentCallback: OnFavoriteSaved? = null
    private val comboNameString: MutableLiveData<String> = MutableLiveData("")
    private lateinit var form: View
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        addToListInFragmentCallback = context as OnFavoriteSaved?

        form = activity!!.layoutInflater.inflate(R.layout.favorites_dialog, null)

        form.combo_name.onChange {
            comboNameString.value = it
        }

        comboNameString.observe(this, Observer {
            if (it.isEmpty()) {
                if (touched) {
                    form.combo_name.error = getString(R.string.no_combo_text)
                }

                positiveButton.isEnabled = false
            } else {
                touched = true
                positiveButton.isEnabled = true
            }

        })

        val builder = AlertDialog.Builder(activity!!)

        return builder
                .setTitle("Save Favorites Combos")
                .setView(form)
                .setPositiveButton(getString(R.string.ok), this)
                .setNegativeButton(getString(R.string.cancel), null)
                .create()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        form.combo_name.requestFocus()
        dialog?.window?.setSoftInputMode(SOFT_INPUT_STATE_VISIBLE)
        return super.onCreateView(inflater, container, savedInstanceState)

    }

    override fun onStart() {
        super.onStart()
        val d = dialog as AlertDialog?
        if (d != null) {
            this.positiveButton = d.getButton(Dialog.BUTTON_POSITIVE)
        }
    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        val comboName = form!!.findViewById<EditText>(R.id.combo_name)
        // saving a new combo
        val savedCombo = SavedCombo.Builder()
                .withName(comboName.text.toString())
                .withSoundPoolParameters(playingSoundIds)
                .withPlaying(true)
                .build()
        addToListInFragmentCallback!!.saved(savedCombo)
    }

    override fun onDismiss(unused: DialogInterface) {
        super.onDismiss(unused)
        Log.d(javaClass.simpleName, "Goodbye!")
    }

    companion object {
        private const val TAG = "SaveToFavoritesDialog"
    }

}